be/src/storage/rowset/beta_rowset_writer.cpp

```diff
@@ -505,107 +616,204 @@ Status BetaRowsetWriter::_flush_src_rssids() {
     return Status::OK();
 }
 
-OLAPStatus BetaRowsetWriter::add_chunk(const vectorized::Chunk& chunk) {
-    if (_segment_writer == nullptr) {
-        _segment_writer = _create_segment_writer();
-    } else if (_segment_writer->estimate_segment_size() >= config::max_segment_file_size ||
-               _segment_writer->num_rows_written() + chunk.num_rows() >= _context.max_rows_per_segment) {
-        RETURN_NOT_OK(_flush_segment_writer(&_segment_writer));
-        _segment_writer = _create_segment_writer();
+VerticalBetaRowsetWriter::VerticalBetaRowsetWriter(const RowsetWriterContext& context) : BetaRowsetWriter(context) {}
+
+VerticalBetaRowsetWriter::~VerticalBetaRowsetWriter() {
+    if (!_already_built) {
+        for (auto& segment_writer : _segment_writers) {
+            segment_writer.reset();
+        }
+
+        for (int i = 0; i < _num_segment; ++i) {
+            auto path = BetaRowset::segment_file_path(_context.rowset_path_prefix, _context.rowset_id, i);
+            // Even if an error is encountered, these files that have not been cleaned up
+            // will be cleaned up by the GC background. So here we only print the error
+            // message when we encounter an error.
+            auto st = _context.env->delete_file(path);
+            LOG_IF(WARNING, !st.ok()) << "Fail to delete file=" << path << ", " << st.to_string();
+        }
     }
-    if (_segment_writer == nullptr) {
-        return OLAP_ERR_INIT_FAILED;
+}
+
+OLAPStatus VerticalBetaRowsetWriter::add_chunk_by_columns(const vectorized::Chunk& chunk,
```

Comment:
IHMO, the name `add_chunk_by_columns` is somehow confusing, how about `add_columns` or `add_selected_columns` or something better?

Reply:
ok